pool/connection_pool.go

```diff
@@ -224,9 +235,12 @@ func (p *ConnectionPool) ConfiguredTimeout(mode Mode) (time.Duration, error) {
 
 // Add adds a new endpoint with the address into the pool. This function
 // adds the endpoint only after successful connection.
-func (p *ConnectionPool) Add(addr string) error {
+func (p *ConnectionPool) Add(ctx context.Context, addr string) error {
 	e := newEndpoint(addr)
 
+	endpointCtx, cancel := context.WithCancel(context.Background())
+	e.cancel = cancel
```

Comment:
It is required to call the `e.cancel()` for all bad cases in the function.

Reply:
Updated. I moved creation of the context, so it creates right before controller call.